### ISSUE TITLE
Docker image should use basedir directory directly

### ISSUE THREAD

**Issue by dhirajsb** [fabric8io/docker-maven-plugin#260]
Using basedir property generates an instruction in generated Dockerfile to CP the contents of /maven to the directory in basedir. Instead it would be more optimal if the plugin generated the target directory in the Dockerfile without having to CP any files around.

**Comment by rhuss** [comment]
That's not quite true: The plugin prepares the data in a maven directory in target but adds the content with a Dockerfile COPY directive directly into the base directory.  There should be no /maven at all in your image.

You can check the generated Dockerfile in `target/docker/<image-name>/build/Dockerfile` which e.g. should look like:

```
FROM busybox:latest
MAINTAINER docker-maven-plugin@jolokia.org
COPY maven /app/
```

Locally it's the `maven` dir (in target) but copied in `/app` when you use `basedir=/app`.

Do you have an example for me where it doesn't work as expected ? I would happily check this out.

**Comment by dhirajsb** [comment]
Maybe that's what it was, I'm not sure I remember correctly now and I could be wrong. Can you try the quickstart https://github.com/fabric8io/quickstarts/tree/master/quickstarts/karaf4/camel-log?

**Comment by dhirajsb** [comment]
I just checked it and it is as you said. Even though, would it be possible to use basedir as the work directory locally instead of a generic /maven?

**Comment by rhuss** [comment]
Yes, this should be no problem and can be easily down because its only an internal implementation details (the Dockerfile itself is a generated, volatile, temporary file ).

The workflow goes as follows:

* The docker-maven-plugin creates the files specified in the assembly descriptor and creates a temporary file name `target/docker/..../build/maven` where all the files are copied into (btw, there is also a buildmode wich doesn't even create this directory but creates a tar file)
* The plugin creates an image at the local docker host (and copies the files over into image's basedir)
* Afterwards the build dir is not used anymore (and gets removed with `mvn clean`)

But I wonder what would be the benefit to use another temporary name which will never appear in any image ?

If you don't mind I will change it to `assembly`, fits properly better its nature.

**Comment by jgangemi** [comment]
you'd also need a special case for when the basedir is `/` so you don't include extraneous files that are also sitting directly under `target`.

**Comment by rhuss** [comment]
@jgangemi sorry, don't get that ;) When basedir is `/` the everything *below* the build dir `target/..../maven` (or whatever ...) will be copied to `/` in the image. In no case anything outside the assembly dir (that local "maven" thing) will be copied into the image.

Or do I completely miss the party here ?

**Comment by jgangemi** [comment]
i probably explained it poorly. :)

if you used the value of `basedir` (let's assume `app`) as the temporary assembly directory, then all the artifacts would be placed under `target/app`. if basedir is `/`, the all artifacts would be placed under `target` which could end up including other things like `classes`, etc which you most likely wouldn't want.

tl;dr: i don't care what we call it but i don't think we should move away from having a specific assembly/temporary dir under `target` that gets used by the plugin to place all artifacts for inclusion.

**Comment by jgangemi** [comment]
actually - we create a `docker` dir under `target` and place related files there, but i think it's still possible to end up doing what i described above if using `/`.

i also just discovered a minor bug w/ how we write create that tmp dir structure.

**Comment by rhuss** [comment]
IMO `basedir` is sth completely different: Its the directory in which artefacts are copied **in the image** no the local temporary assembly directory. This local temporary directory is always "maven" (and is in `target/docker/<image name>/build`) and is fixed. 

The Dockerfile then contains 

COPY maven/  basedir

That's the purpose of basedir.

**Comment by jgangemi** [comment]
yeah - we're on the same page w/ that.

**Comment by rhuss** [comment]
> if you used the value of basedir (let's assume app) as the temporary assembly directory, then all the artifacts would be placed under target/app. if basedir is /, the all artifacts would be placed under target which could end up including other things like classes, etc which you most likely wouldn't want.

ok, so that we agree that this ^^ can't happen. The assembly dir will be always constant. If basedir is '/' the everything will copied alll down root within the image.

**Comment by rhuss** [comment]
@dhirajsb : Could you please give an use case where a different temporary local directory name other than `maven` would be a be a benefit ? (see https://github.com/rhuss/docker-maven-plugin/issues/260#issuecomment-130818385).

I can change that easily to a more maven-agnostic name (it's an implementation detail) but I wonder what it would be good for.

**Comment by crossdot** [comment]
I need change output directory too. I've got an external Dockerfile with image building specifications, and there already some files under ./webapps directory, then I want to assembly maven artifact there, but after docker:build there no any "COPY maven basedir" in Dockerfile. Maybe it's a bug.
I don't understand why not just copy .war to my custom webapps directory without additional specifications in Dockerfile?

**Comment by kifj** [comment]
Please update at least the documentation - when using an external Dockerfile you must include "COPY maven basedir" in your Dockerfile, and the description of assembly and basedir is misleading. 

With an external Dockerfile there are some benefits if basedir can be set to the webapps folder of tomcat, this saves you the COPY line in the Dockerfile. 

And it would have saved me some time.

**Comment by dhirajsb** [comment]
Closing stale issue
